README.md

```diff
@@ -256,19 +281,24 @@ immediate.
 ### Creating strings
 
 ```
-(string.new $memory $encoding ptr:address len:i32)
+(string.new_wtf8 $memory ptr:address bytes:i32)
   -> str:stringref
 ```
-Create a new string from the *`len`* bytes encoded in memory at *`ptr`*.
-Out-of-bounds access will trap.  The `utf-16` encoding implies two-byte
-alignment for both operands and will trap otherwise.  Attempting to
-create a string with bytes that are not valid for the encoding will
-trap.  No terminating `NUL` byte is expected.
+Create a new string from the *`bytes`* WTF-8 bytes in memory at *`ptr`*.
+Out-of-bounds access will trap.  Attempting to create a string with
+invalid WTF-8 will trap.  The maximum value for *`bytes`* is
+2<sup>31</sup>–1; passing a higher value traps.
```

Comment:
Standardizing the max string length (in particular: in the Wasm-JS API spec) might be challenging. In V8, the current situation is that we allow up to `(1 << 28) - 16` characters on 32-bit builds and up to `(1 << 29) - 24` characters on 64-bit builds. The latter cannot be increased without massive implementation effort; I'm not sure whether the former could be increased to match the latter. I wouldn't be surprised if other browsers similarly had hard constraints defined by their implementation choices. The Wasm-JS API so far has been able to standardize precise, implementation-independent limits for everything -- but so far, no Wasm feature required this kind of tight alignment with existing implementations of JS features and their limits.

Reply:
Related investigation into string lengths: https://github.com/wingo/stringrefs/issues/12#issuecomment-951988797.  I was very surprised that JSC allows strings up to 2^31-1 code units!!  But OK.

I think the intention here is to allow WebAssembly to address any string that comes from the host.  However when it comes to creating strings, unlike pre-GC WebAssembly, there is the possibility of dynamic allocation failure.  *Any* `string.new` can fail, regardless of the length.  This brings an inherent source of non-uniformity into the execution of a WebAssembly program but seems to be essential to the domain.  On the other hand dynamic allocation failure can provide cover for implementation-specific limits; see also https://github.com/wingo/stringrefs#stringnew-size-limits.  WDYT?

Reply:
Well, it's not technically all the same: if an attempt to allocate a (small) string fails due to OOM, we will crash the process (not trap). If an allocation attempt fails because of a (spec or implementation defined) limit, we would at least have the option to trap instead. But I don't have a better suggestion that just relying on the generic "any allocation can fail" rule here.